### DISCUSSION TITLE
Why does `Concatenate` only work one way?

### DISCUSSION THREAD

**Discussion by decorator-factory** [microsoft/pyright#2167] [Q&A] [answered]
What's the rationale behind `Concatenate` (from PEP 612) only allowing a `ParamSpec` variable to appear on the right?
i.e. `Concatenate[int, P]` is legal, but `Concatenate[P, int]` is not.

I wanted to make a decorator that added authentication and authorization handling on HTTP controllers:
```py
@requires("admin")
async def delete_everything(req: Request, user: AuthenticatedUser) -> Response:
    ... 
# delete_everything becomes:
# async (req: Request) -> Response
# which can be then used as a standard handler in framework, in my case Starlette
```
 Because I wanted it to work on methods as well as with other similar decorators:
```py
@requires("admin")
@parse_body_as(UserMention)
def ban_user(self, mention: UserMention, user: AuthenticatedUser) -> Response:
    ...
# ban_user becomes:
# async (self, req: Request) -> Response
# and then I can use the bound method as a handler
```
Traditionally things like this have been done without adding arguments, but with adding extra attributes to the `Request` objects, but that's not possible to express in type hints as far as I understand.

I am aware that this is just how the grammar is defined in the PEP, but maybe there's some deeper reason why what I wanted isn't possible?

This is possible in TypeScript, so I was surprised it doesn't work with PEP 612.

**Comment by layday** [comment, marked as answer]
It's because `Callable` only supports expressing positional arguments. If you have `Callable[Concatenate[P, int], None]`, then what should `P` capture?  It can't capture `kwargs` so that's a limitation that'd have to be imposed on `ParamSpec` in this context.  But also, if you imagine this as `def foo(*args, bar: int)`, what that means is the `int` would have to be a keyword-only argument, and keyword-only arguments are not supported by `Callable`.
